siliconcompiler/tools/vpr/place.py

```diff
@@ -38,7 +41,25 @@ def pre_process(chip):
     step = chip.get('arg', 'step')
     index = chip.get('arg', 'index')
 
-    if not chip.valid('input', 'constraint', 'pins', default_valid=True):
+    part_name = chip.get('fpga', 'partname')
+
+    if 'pinmap' in chip.getkeys('input', 'constraint'):
+
+        constraint_file = vpr.find_single_file(chip, 'input', 'constraint', 'pinmap',
+                                               step=step, index=index,
+                                               file_not_found_msg="JSON constraints file not found")
+
+        map_file = vpr.find_single_file(chip, 'fpga', part_name, 'file', 'constraints_map',
+                                        file_not_found_msg="constraints map not found")
+
+        constraints_map = load_constraints_map(map_file)
+        json_constraints = load_json_constraints(constraint_file)
+        all_place_constraints = map_constraints(json_constraints, constraints_map)
+
+        constraints_xml = generate_vpr_constraints_xml(all_place_constraints)
+        write_vpr_constraints_xml_file(constraints_xml, vpr.auto_constraints())
+
+    elif not chip.valid('input', 'constraint', 'pins', default_valid=True):
```

Comment:
why default_valid?

Reply:
Unclear at the moment; everything on that line except 'el' is from a previous PR.  Reviewing SC docs makes my choice to use that parameter setting suspect.  Will try deleting and retesting and revisit only if tests fail.

Reply:
Removing default_valid=True causes the following failures; all tests pass otherwise.  Based on which pass (the three constraints tests), I believe we added default_valid=True to allow the flow to keep working when no constraints are given?

```
FAILED test_dspblackbox_fpgaflow.py::test_dspblackbox_fpgaflow - siliconcompiler._common.SiliconCompilerError: These final steps could not be reached: ['route']
FAILED test_dspextract_fpgaflow.py::test_dspextract_fpgaflow - siliconcompiler._common.SiliconCompilerError: These final steps could not be reached: ['route']
FAILED test_dspmap_fpgaflow.py::test_dspmap_fpgaflow - siliconcompiler._common.SiliconCompilerError: These final steps could not be reached: ['route']
FAILED test_flopmap_fpgaflow.py::test_flopmap_fpgaflow - siliconcompiler._common.SiliconCompilerError: These final steps could not be reached: ['route']
FAILED test_fpgaflow_apr.py::test_fpgaflow_apr - siliconcompiler._common.SiliconCompilerError: These final steps could not be reached: ['route']
FAILED test_fpgaflow.py::test_fpgaflow - siliconcompiler._common.SiliconCompilerError: These final steps could not be reached: ['bitstream']
FAILED test_matrix_multiply_fpgaflow.py::test_matrix_multiply_fpgaflow - siliconcompiler._common.SiliconCompilerError: These final steps could not be reached: ['route']
FAILED test_mem_to_flops_fpgaflow.py::test_mem_to_flops_fpgaflow - siliconcompiler._common.SiliconCompilerError: These final steps could not be reached: ['route']
FAILED test_vhdl_fpgaflow.py::test_fpgaflow - siliconcompiler._common.SiliconCompilerError: These final steps could not be reached: ['bitstream']
```

Reply:
I understand this issue now; the default_valid=True was indeed masking a bug in the setup.  Adding a hook in place.py's `pre_process()` function to block constraints file generation when no constraints are provided resolves this.